### ISSUE TITLE
The delay in the store's update confuses react-ace

### ISSUE THREAD

**Issue by SidneyNemzer** [tshaddix/react-chrome-redux#97]
I've got react-ace running in a devtools panel. There's a redux store in the background page, connected to the devtools with react-chrome-redux.

Typing in the editor works! It syncs the content across any number of devtools panels, and can save to Chrome's storage for session persistence.

However, it seems I've run into an issue that's caused by the slight delay react-chrome-redux introduces: when inserting a snippet from Ace, the cursor jumps to the **beginning** of the snippet, instead of the end.

I've simulated the delay with a regular redux store by wrapping the redux store's `dispatch()` function with a `setTimeout` of 2ms, and the same issue occurs. When using a vanilla redux store with no delay, autocomplete works correctly (that is, the cursor is positioned at the end of the inserted code).

Any ideas on how I'd fix this? If the code or a video demonstrating the issue would help, I can add them.

**Comment by vhmth** [comment]
Hi Sidney!

Do you have some code I can look at? That would be really helpful so I can point you to how you could solve this issue.

**Comment by SidneyNemzer** [comment]
Sorry for the delay, @vhmth

I was about to point you to the project that I first noticed this issue in, but I thought a [mcve](https://stackoverflow.com/help/mcve) would be more helpful.

So here it is: https://github.com/SidneyNemzer/delay-demo

The readme has basic instructions for how to install/view.

**Note:* I was able to come up with a fix to the issue, although I'm not sure it's ideal. The demo shows the broken version, and the fixed version. I've released my actual project using this fix.

If you *do* want to see the project that I'm referring to: https://github.com/SidneyNemzer/snippets

**Comment by SidneyNemzer** [comment]
Update: I've had similar issues with Ace and [Elm](http://elm-lang.org/). I think that updating Ace's value when the store (Redux) or model (Elm) updates isn't very preformant, and it causes issues with cursor position, autocomplete, etc.

Ideally, Ace would only be notified of an update when it actually needs to change it's content. For example, when switching tabs to a different piece of code. Otherwise, Ace would be allowed to hold its own internal state and notify the surrounding code of changes the user has made.

**Comment by vhmth** [comment]
Hey @SidneyNemzer sorry for the late reply here. This is the problem with being a co-founder of a company and trying to also keep on top of your OS game 😆.

https://www.useloom.com/share/b47cf5f04dc4481d83c17380bd5d4103

**Comment by SidneyNemzer** [comment]
First, sorry for the weird indenting... I managed to mix tabs and spaces somehow. I might fix that at some point.

You said that I need to wait for the store to initialize, although I don't think that's an issue here because the problem happens on store *updates*, not when the store is created.

In [the "test" file](https://github.com/SidneyNemzer/delay-demo/blob/551070a85cb128a015877ed68f4bed5e0c8afb83/src/js/test.js), the problem is reproduced by forcing store updates to run on the next tick, like react-chrome-redux does. In fact, that example doesn't even use a react-chrome-redux store, just a regular redux store (and some special logic to intercept dispatches).

I think this confuses Ace because after Ace notifies Redux of changes to the file, Redux has to propagate the update to the store. Then React is notified of the change, and re-renders Ace, setting the content to the new value. But Ace is like "Oh, I need to set a new value? OK, got it" and it acts like a new document is being displayed rather than the old one having been updated.

If that all happens in the same tick, I guess Ace doesn't treat it like a new document. So *usually* React and Ace play nicely. But when React is delayed in re-rendering Ace, Ace has to assume it's a new document. I had this same problem when using Ace and Elm, and I worked around it by only setting Ace's content when the user selected a new document, instead of on every re-render. A similar fix could be used with React.

In conclusion, this seems to be a problem between React and Ace, and I don't think react-chrome-redux can prevent it. Because of that, I think I'll close this issue.

I appreciate that you were able to take some time to look at this though. I don't think I've ever gotten a video reply to one of my issues ;)

**Comment by vhmth** [comment]
Oooooh I see. I'm just now noticing what the `Proxy` is doing and your comment `// Even a delay of 0 causes an issue...`. That is very strange. I'm not sure what's going on with Ace in that case. Let me know here once you figure it out/if it's a bug on their end!

Good luck!

**Comment by SidneyNemzer** [comment]
I finally circled back to this, it turns out the issue is that webext-redux makes redux store updates asynchronous, which allows Ace to process new keyboard events before the update has come back from the background page.

I detailed the issue (and came up with some potential solutions) in this document: https://github.com/SidneyNemzer/snippets/blob/master/docs/async-store-workaround.md

Basically, webext-redux can't really prevent this issue without some major changes. I worked around the issue by using an uncontrolled component (more details in the document linked above).
